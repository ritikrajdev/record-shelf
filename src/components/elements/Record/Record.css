.record {
  --height: 300px;

  width: 300px;
  height: var(--height);
  position: relative;
  background-color: var(--dark-gray);
  border-radius: 8px;
  box-shadow: 4px 4px 8px #00000020;
  padding: 16px;
}

.record:nth-child(2n - 1) {
  background-color: var(--light-gray);
}

.record-img-container {
  height: 75%;
  position: relative;
  text-align: center;
}

.record-img {
  height: calc(var(--height) * 0.75 - 1rem);
  box-sizing: border-box;
  margin: 0.5rem;
  border-radius: 50%;
  border: 1px solid black;
  box-sizing: border-box;
}

.record-info {
  display: flex;
  justify-content: space-between;
  gap: 1rem;
  align-items: center;
  margin-top: 1rem;
}

.record-info > div {
  max-width: calc(100% - 4rem);
}

.record-name {
  overflow: hidden;
  text-overflow: ellipsis;
  -webkit-line-clamp: 1;
  height: 1.3rem;
  width: 100%;
  font-size: 1.2rem;
  font-weight: 200;
}
