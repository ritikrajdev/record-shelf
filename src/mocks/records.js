export const mockRecordsData = [
  {
    id: '0e2ddccf-f494-4c41-a499-90f8267f491a',
    name: 'Beautiful Mistakes (feat. Megan Thee Stallion)',
    genre: 'Pop',
    artist: 'Maroon 5',
    imageUrl:
      'https://i.scdn.co/image/ab67616d0000b27386a8ab515de4b7aef28cd631',
    publishedAt: '2021-06-11T00:00:00',
    numLikes: 0,
    isLiked: false,
  },
  {
    id: '45e1d753-2986-43cb-9b9d-30c370056319',
    name: 'This Love',
    genre: 'Pop',
    artist: 'Maroon 5',
    imageUrl:
      'https://i.scdn.co/image/ab67616d0000b27317b3850d758fff5a2301e537',
    publishedAt: '2002-06-25T00:00:00',
    numLikes: 2,
    isLiked: true,
  },
  {
    id: '8266dac3-b2b1-4551-b2ec-91e260244a51',
    name: 'Payphone',
    genre: 'Pop',
    artist: 'Maroon 5',
    imageUrl:
      'https://i.scdn.co/image/ab67616d0000b2733119f490f02fcee6514e8604',
    publishedAt: '2012-01-01T00:00:00',
    numLikes: 2,
    isLiked: true,
  },
];
