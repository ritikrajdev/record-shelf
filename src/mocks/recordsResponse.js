export const mockRecordsGetResponse = {
  data: [
    {
      id: '0e2ddccf-f494-4c41-a499-90f8267f491a',
      name: 'Beautiful Mistakes (feat. Megan Thee Stallion)',
      genre: {
        id: '128aa7f8-c943-48ce-b352-7edd26fa4c6e',
        name: 'Pop',
      },
      artist: {
        id: '496b0a85-2bfa-45bc-8d0f-57fe0ce55708',
        name: 'Maroon 5',
      },
      imageUrl:
        'https://i.scdn.co/image/ab67616d0000b27386a8ab515de4b7aef28cd631',
      publishedAt: '2021-06-11T00:00:00',
    },
    {
      id: '45e1d753-2986-43cb-9b9d-30c370056319',
      name: 'This Love',
      genre: {
        id: '128aa7f8-c943-48ce-b352-7edd26fa4c6e',
        name: 'Pop',
      },
      artist: {
        id: '496b0a85-2bfa-45bc-8d0f-57fe0ce55708',
        name: 'Maroon 5',
      },
      imageUrl:
        'https://i.scdn.co/image/ab67616d0000b27317b3850d758fff5a2301e537',
      publishedAt: '2002-06-25T00:00:00',
    },
    {
      id: '8266dac3-b2b1-4551-b2ec-91e260244a51',
      name: 'Payphone',
      genre: {
        id: '128aa7f8-c943-48ce-b352-7edd26fa4c6e',
        name: 'Pop',
      },
      artist: {
        id: '496b0a85-2bfa-45bc-8d0f-57fe0ce55708',
        name: 'Maroon 5',
      },
      imageUrl:
        'https://i.scdn.co/image/ab67616d0000b2733119f490f02fcee6514e8604',
      publishedAt: '2012-01-01T00:00:00',
    },
    {
      id: '94af94a3-9d47-4dd8-9b45-1c51a9443e96',
      name: 'Maps',
      genre: {
        id: '128aa7f8-c943-48ce-b352-7edd26fa4c6e',
        name: 'Pop',
      },
      artist: {
        id: '496b0a85-2bfa-45bc-8d0f-57fe0ce55708',
        name: 'Maroon 5',
      },
      imageUrl:
        'https://i.scdn.co/image/ab67616d0000b273442b53773d50e1b5369bb16c',
      publishedAt: '2014-09-02T00:00:00',
    },
    {
      id: 'cd3cc1e3-e1e0-4ccd-bc67-747648985838',
      name: 'Lost',
      genre: {
        id: '128aa7f8-c943-48ce-b352-7edd26fa4c6e',
        name: 'Pop',
      },
      artist: {
        id: '496b0a85-2bfa-45bc-8d0f-57fe0ce55708',
        name: 'Maroon 5',
      },
      imageUrl:
        'https://i.scdn.co/image/ab67616d0000b27386a8ab515de4b7aef28cd631',
      publishedAt: '2021-06-11T00:00:00',
    },
    {
      id: '0530d74c-5dbc-4f1b-b3e2-7ada4520fb4c',
      name: 'Magnetic',
      genre: {
        id: '305202d2-b598-4c1b-a2bb-ca08bb65bad3',
        name: 'Country',
      },
      artist: {
        id: '8150acae-62fd-491e-bac4-06d074b3c7b4',
        name: 'Phillip Phillips',
      },
      imageUrl:
        'https://i.scdn.co/image/ab67616d0000b27322d30dee7738733b99774f9b',
      publishedAt: '2018-01-19T00:00:00',
    },
  ],
};

export const mockLikeGetResponse = {
  data: {
    like: false,
    count: 0,
  },
};

export const mockLikePatchResponse = {
  data: {
    like: true,
    count: 1,
  },
};
