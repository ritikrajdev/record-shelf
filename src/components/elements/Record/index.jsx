import PropTypes from 'prop-types';
import React from 'react';
import IconButton from '../IconButton';

import './Record.css';

import heartGray from '../../../assets/icons/heart-gray.svg';
import heartRed from '../../../assets/icons/heart-red.svg';

export default function Record({ recordData, onLikeClick }) {
  return (
    <div className='record'>
      <div className='record-img-container'>
        <img className='record-img' src={recordData.imageUrl} alt='image' />
      </div>
      <div className='record-info'>
        <div>
          <div className='record-name'>{recordData.name}</div>
          <div className='record-artist'>{recordData.artist}</div>
        </div>

        <IconButton
          color='white'
          iconSrc={recordData.isLiked ? heartRed : heartGray}
          width='3rem'
          height='3rem'
          onClick={onLikeClick}
        >
          {recordData.numLikes}
        </IconButton>
      </div>
    </div>
  );
}

Record.propTypes = {
  recordData: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    artist: PropTypes.string.isRequired,
    imageUrl: PropTypes.string.isRequired,
    publishedAt: PropTypes.string.isRequired,
    numLikes: PropTypes.number.isRequired,
    isLiked: PropTypes.bool.isRequired,
  }),
  onLikeClick: PropTypes.func,
};
