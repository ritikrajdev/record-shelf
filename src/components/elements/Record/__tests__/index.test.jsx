import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import Record from '..';
import { mockRecordsData } from '../../../../mocks/records';

describe('Record', () => {
  it('should render correctly with gray heart when not liked', () => {
    const screen = render(<Record recordData={mockRecordsData[0]} />);
    expect(screen.asFragment()).toMatchSnapshot();
    expect(screen.getByRole('button').style.background).toContain('heart-gray');
  });

  it('should render red heart when liked', () => {
    const screen = render(
      <Record recordData={{ ...mockRecordsData[1], isLiked: true }} />
    );
    expect(screen.getByRole('button').style.background).toContain('heart-red');
  });

  it('should call onClick when like button is clicked', () => {
    const onClick = jest.fn();
    const screen = render(
      <Record recordData={mockRecordsData[0]} onLikeClick={onClick} />
    );
    fireEvent.click(screen.getByRole('button'));
    expect(onClick).toHaveBeenCalled();
  });
});
