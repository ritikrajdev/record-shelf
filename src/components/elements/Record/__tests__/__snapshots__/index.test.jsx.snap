// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Record should render correctly with gray heart when not liked 1`] = `
<DocumentFragment>
  <div
    class="record"
  >
    <div
      class="record-img-container"
    >
      <img
        alt="image"
        class="record-img"
        src="https://i.scdn.co/image/ab67616d0000b27386a8ab515de4b7aef28cd631"
      />
    </div>
    <div
      class="record-info"
    >
      <div>
        <div
          class="record-name"
        >
          Beautiful Mistakes (feat. Megan Thee Stallion)
        </div>
        <div
          class="record-artist"
        >
          Maroon 5
        </div>
      </div>
      <button
        class="icon-button"
        style="background-color: transparent; background-image: url(heart-gray.svg); background-repeat: no-repeat; background-position: center; background-size: contain; width: 3rem; height: 3rem; color: white;"
      >
        0
      </button>
    </div>
  </div>
</DocumentFragment>
`;
